dialogue-core/src/main/java/com/palantir/dialogue/core/ChannelToEndpointChannel.java

```diff
@@ -16,23 +16,21 @@
 
 package com.palantir.dialogue.core;
 
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
 import com.google.common.util.concurrent.ListenableFuture;
 import com.palantir.dialogue.Channel;
 import com.palantir.dialogue.Endpoint;
 import com.palantir.dialogue.Request;
 import com.palantir.dialogue.Response;
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.function.Function;
+import java.util.function.BiFunction;
 
 final class ChannelToEndpointChannel implements Channel {
 
-    private final Function<Endpoint, Channel> adapter;
-    private final Map<Object, Channel> cache;
+    private final LoadingCache<Endpoint, Channel> cache;
 
-    ChannelToEndpointChannel(Function<Endpoint, Channel> adapter) {
-        this.adapter = adapter;
-        this.cache = new ConcurrentHashMap<>();
+    ChannelToEndpointChannel(Channel channel, BiFunction<Channel, Endpoint, Channel> loader) {
```

Comment:
I tweaked the loader here to take the delegate channel as an argument rather than capturing it. This makes the usage a bit cleaner by avoiding the need to introduce a final `tracingChannel` variable for capturing.

Reply:
I think I'd prefer to revert this piece -- it's nice to avoid the capture, but it makes it harder to understand what the function represents within this class. The fact the original channel may be returned is an implementation detail of a function passed to this class.

Reply:
Yeah this makes sense. I can accomplish what I was trying to here without this change.